sql/parser/eval.go

```diff
@@ -319,109 +319,236 @@ type cmpArgs struct {
+		},
 	},
-	cmpArgs{LE, bytesType, bytesType}: func(left Datum, right Datum) (DBool, error) {
-		return DBool(left.(DBytes) <= right.(DBytes)), nil
+	cmpArgs{LE, bytesType, bytesType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return DBool(left.(DBytes) <= right.(DBytes)), nil
+		},
 	},
-	cmpArgs{LE, boolType, boolType}: func(left Datum, right Datum) (DBool, error) {
-		return DBool(!left.(DBool) || right.(DBool)), nil
+	cmpArgs{LE, boolType, boolType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return DBool(!left.(DBool) || right.(DBool)), nil
+		},
 	},
-	cmpArgs{LE, intType, intType}: func(left Datum, right Datum) (DBool, error) {
-		return DBool(left.(DInt) <= right.(DInt)), nil
+	cmpArgs{LE, intType, intType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return DBool(left.(DInt) <= right.(DInt)), nil
+		},
 	},
-	cmpArgs{LE, floatType, floatType}: func(left Datum, right Datum) (DBool, error) {
-		return DBool(left.(DFloat) <= right.(DFloat)), nil
+	cmpArgs{LE, floatType, floatType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return DBool(left.(DFloat) <= right.(DFloat)), nil
+		},
 	},
-	cmpArgs{LE, dateType, dateType}: func(left Datum, right Datum) (DBool, error) {
-		return DBool(right.(DDate).Before(left.(DDate).Time)), nil
+	cmpArgs{LE, dateType, dateType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return DBool(right.(DDate).Before(left.(DDate).Time)), nil
+		},
 	},
-	cmpArgs{LE, timestampType, timestampType}: func(left Datum, right Datum) (DBool, error) {
-		return !DBool(right.(DTimestamp).Before(left.(DTimestamp).Time)), nil
+	cmpArgs{LE, timestampType, timestampType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return !DBool(right.(DTimestamp).Before(left.(DTimestamp).Time)), nil
+		},
 	},
-	cmpArgs{LE, intervalType, intervalType}: func(left Datum, right Datum) (DBool, error) {
-		return DBool(left.(DInterval).Duration <= right.(DInterval).Duration), nil
+	cmpArgs{LE, intervalType, intervalType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			return DBool(left.(DInterval).Duration <= right.(DInterval).Duration), nil
+		},
 	},
 
-	cmpArgs{Like, stringType, stringType}: func(left Datum, right Datum) (DBool, error) {
-		pattern := regexp.QuoteMeta(string(right.(DString)))
-		// Replace LIKE specific wildcards with standard wildcards
-		pattern = strings.Replace(pattern, "%", ".*", -1)
-		pattern = strings.Replace(pattern, "_", ".", -1)
-		pattern = anchorPattern(pattern, false)
+	cmpArgs{Like, stringType, stringType}: {
+		fn: func(left Datum, right Datum, cache cmpOpCacheVal) (DBool, error) {
+			re := cache.(*regexp.Regexp)
+			return DBool(re.MatchString(string(left.(DString)))), nil
+		},
+		initFn: func(left Datum, right Datum) (cmpOpCacheVal, error) {
```

Comment:
Having a separate `initFn` like this is nice and explicit. I was initially thinking that we'd have only a single function that took a `cache *interface{}` arg which the function could check to see if the `cache` was populated. Something like:

```
  func(left Datum, right Datum, cache *interface{}) (DBool, error) {
    var re *regexp.Regep
    switch t := (*cache).(type) {
    case *regexp.Regexp:
       re = t
    default:
       pattern := ...
       re = regexp.MustCompile(pattern, nil)
       *cache = re
    }
    return DBool(re.MatchString(string(left.(DString)))), nil
  }
```

Taking this approach would be less overall change. I'm not sure if there are other reasons to choose one approach over another.